In [28]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

Airbnb = pd.read_csv(r"C:\Users\sebas\OneDrive\Dokumente\Schule\Informatik\M259\AB_NYC_2019.csv") 
Airbnb.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  10/19/2018   
1  Entire home/apt    225               1                 45   5/21/2019   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270    7/5/2019   
4  Entire home/apt     80              10                  9  11/19/2018   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [29]:
#3.1

X = Airbnb[["latitude",
    "longitude",
    "minimum_nights",
    "number_of_reviews",
    "reviews_per_month",
    "calculated_host_listings_count",
    "availability_365"]]  

y = Airbnb["price"]  

X = X.fillna(0)
y = y.fillna(0)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [30]:
#3.2
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

## 3.2
Für die Vorhersage des Preises eines Airbnb Angebots habe ich die Lineare Regression aus sklearn gewählt.
Dieser Algorithmus passt gut, da der Preis ein kontinuierliches Ziel ist und die Beziehung zwischen Preis und Merkmalen (wie Lage, Mindestnächte oder Anzahl Bewertungen) oft linear oder fast linear ist.
Lineare Regression ist einfach, schnell zu trainieren und liefert oft gute erste Ergebnisse.
Ausserdem ist das Modell leicht verständlich, was bei der ersten Analyse hilft.

In [31]:
#3.3

# Vorhersagen für den Testdatensatz
y_pred = model.predict(X_test)

# Vorhersagen und Werte vergleichen
for i in range(20):
    print(f"Beispiel {i+1}: Vorhergesagt = {y_pred[i]:.2f}, Tatsächlich = {y_test.iloc[i]}")

Beispiel 1: Vorhergesagt = 161.78, Tatsächlich = 89
Beispiel 2: Vorhergesagt = 114.78, Tatsächlich = 30
Beispiel 3: Vorhergesagt = 173.00, Tatsächlich = 120
Beispiel 4: Vorhergesagt = 249.26, Tatsächlich = 470
Beispiel 5: Vorhergesagt = 134.45, Tatsächlich = 199
Beispiel 6: Vorhergesagt = 98.28, Tatsächlich = 170
Beispiel 7: Vorhergesagt = 144.14, Tatsächlich = 90
Beispiel 8: Vorhergesagt = 86.41, Tatsächlich = 60
Beispiel 9: Vorhergesagt = 159.28, Tatsächlich = 150
Beispiel 10: Vorhergesagt = 51.59, Tatsächlich = 74
Beispiel 11: Vorhergesagt = 144.47, Tatsächlich = 28
Beispiel 12: Vorhergesagt = 165.50, Tatsächlich = 200
Beispiel 13: Vorhergesagt = 157.70, Tatsächlich = 230
Beispiel 14: Vorhergesagt = 144.80, Tatsächlich = 48
Beispiel 15: Vorhergesagt = 150.60, Tatsächlich = 100
Beispiel 16: Vorhergesagt = 165.92, Tatsächlich = 100
Beispiel 17: Vorhergesagt = 146.74, Tatsächlich = 115
Beispiel 18: Vorhergesagt = 181.20, Tatsächlich = 94
Beispiel 19: Vorhergesagt = 134.92, Tatsächlich 

## 3.3
Die Vorhersagen des Modells weichen teilweise von den tatsächlichen Werten ab.
In einigen Fällen, wie Beispiel 1 (Vorhersage: 161,78 – tatsächlich: 89) und Beispiel 4 (Vorhersage: 249,26 – tatsächlich: 470), sind die Unterschiede besonders gross.
Das deutet darauf hin, dass das Modell bei sehr niedrigen oder sehr hohen Preisen Probleme hat.
Weitere Anpassungen oder zusätzliche Features könnten helfen, die Abweichungen zu verringern und die Genauigkeit zu verbessern.